handler/handler.go

```diff
@@ -0,0 +1,307 @@
+	"online/middleware"
+	"online/models"
+	"online/repository"
+
+	//Inbuild packages
+	"fmt"
+	"net/http"
+	"reflect"
+	"regexp"
+
+	//Third party packages
+	"github.com/fatih/structs"
+	"github.com/labstack/echo"
+	"golang.org/x/crypto/bcrypt"
+	"gorm.io/gorm"
+)
+
+type Database struct {
+	Db *gorm.DB
+}
+
+// This is for Signup
+func (db Database) Signup(c echo.Context) error {
+	var (
+		data models.User
+		role models.Roles
+	)
+	log := logs.Log()
+	log.Info.Println("Message : 'signup-API called'")
+
+	//Get user details from request body
+	if err := c.Bind(&data); err != nil {
+		log.Error.Println("Error : 'internal server error' Status : 500")
+		return c.JSON(http.StatusInternalServerError, map[string]interface{}{
+			"status": 500,
+			"error":  "internal server error",
+		})
+	}
+
+	//To check if any credential is missing or not
+	fields := structs.Names(&models.SignupReq{})
+	for _, field := range fields {
+		if reflect.ValueOf(&data).Elem().FieldByName(field).Interface() == "" {
+			stmt := fmt.Sprintf("missing %s", field)
+			log.Error.Printf("Error : '%s' Status : 400\n", stmt)
+			return c.JSON(http.StatusBadRequest, map[string]interface{}{
+				"status": 400,
+				"error":  stmt,
+			})
+		}
+	}
+
+	//validate email format
+	emailRegex := regexp.MustCompile(`^[a-z0-9._%+\-]+@[a-z0-9.\-]+\.[a-z]{2,4}$`)
+	if !emailRegex.MatchString(data.Email) {
+		log.Error.Println("Error : 'Invalid Email' Status : 400")
+		return c.JSON(http.StatusBadRequest, map[string]interface{}{
+			"status": 400,
+			"error":  "Invalid Email",
+		})
+	}
+
+	//validate the password
+	if len(data.Password) < 8 {
+		log.Error.Println("Error : 'password must be greater than 8 characters' Status : 400")
+		return c.JSON(http.StatusBadRequest, map[string]interface{}{
+			"status": 400,
+			"error":  "password must be greater than 8 characters",
+		})
+	}
+
+	//validate the role
+	if data.Role != "admin" && data.Role != "user" {
+		log.Error.Println("Error : 'Invalid role' Status : 400")
+		return c.JSON(http.StatusBadRequest, map[string]interface{}{
+			"status": 400,
+			"error":  "Invalid role",
+		})
+	}
+
+	//To check if the user details already exist or not
+	data, err := repository.ReadUserByEmail(db.Db, data)
+	if err == nil {
+		log.Error.Println("Error : 'user already exist' Status : 409")
+		return c.JSON(http.StatusBadRequest, map[string]interface{}{
+			"status": 409,
+			"error":  "user already exist",
+		})
+	}
+
+	//To change the password into hashedPassword
+	password, err := bcrypt.GenerateFromPassword([]byte(data.Password), bcrypt.DefaultCost)
+	if err != nil {
+		log.Error.Printf("Error : '%s'\n", err)
+		return nil
+	}
+	data.Password = string(password)
+
+	//Select a role_id for specified role
+	role, _ = repository.ReadRoleIdByRole(db.Db, data)
+	data.RoleId = role.RoleId
+
+	//Adding a user details into our database
+	if err = repository.CreateUser(db.Db, data); err != nil {
+		log.Error.Println("Error : 'email already exist' Status : 409")
```

Comment:
What is meant by "409" status code?

Reply:
"409" status code indicates the user already exist error. Here also we will get a similar error so that I've used "409" status code.

Reply:
No